luv callbacks: throw error on deferred methods instead of crashing

bfredl opened the pull request:
Unfortunately "throw error" in luv callbacks currently means putting garbage on the screen. This is hardcoded in luv, we could use a registered hook that luv should call in case of callback errors instead of writing to stderr.

Test (compare master to PR): 
```
timer = vim.loop.new_timer()
timer:start(1000,0, function ()
  vim.api.nvim_command("sleep 100m")
  vim.api.nvim_command("echo 'test'")
end)
```

Replacement without error:
```
timer = vim.loop.new_timer()
timer:start(1000,0, vim.schedule_wrap(function ()
  vim.api.nvim_command("sleep 100m")
  vim.api.nvim_command("echo 'test'")
end))
```

NB:  `sleep` inside callback doesn't really make sense, it is only used to provoke the crash in a simple and deterministic way.

Fixes #10298 

I also think we should rename `NVIM_API_ASYNC` to `NVIM_API_UNDEFERRED`, to better describe what the attribute actually does. (a bikeshed option might be ASYNC_SAFE, but it still suffers some of the ambiguity of ASYNC, i e deferred methods are definitely "async safe" in the sense of an async notification, it might just get delayed a bit compared to undeferred ones).

bfredl committed:
libluv: use luv_set_callback to control callback execution

Disable the use of deferred API functions in a fast lua callback
Correctly display error messages from a fast lua callback

bfredl committed:
rename: FUNC_API_ASYNC => FUNC_API_FAST

justinmk commented:
> I also think we should rename `NVIM_API_ASYNC` to `NVIM_API_UNDEFERRED`, to better describe what the attribute actually does. (a bikeshed option might be ASYNC_SAFE, but it still suffers some of the ambiguity of ASYNC, i e deferred methods are definitely "async safe" in the sense of an async notification, it might just get delayed a bit compared to undeferred ones).

"undeferred" is sort of  double-negative. We already call the events "fast events", how about `NVIM_API_FAST` (or `NVIM_API_IMMEDIATE`, but "fast" is shorter and consistent with existing jargon) ?

bfredl commented:
Seems we can do this more robustly with a simple luv patch to control the execution of callbacks. https://github.com/bfredl/luv/commit/4a4ff44cc57c148e45a48da35283a4a2f16c6250 . I will ask upstream what they think of it.

bfredl committed:
make vim.loop == require'luv'

This avoids initializing libluv a second time if a plugin invokes
require'luv'. It is probably not an issue, but better to be safe.

justinmk commented on src/nvim/lua/executor.c:
```diff
@@ -546,6 +611,13 @@ Object executor_exec_lua_cb(LuaRef ref, const char *name, Array args,
 
+/// check if the current execution context is safe for calling deferred API
+/// methods. Luv callbacks are unsafe as they are called inside the uv loop.
+bool nlua_check_safe_cb(lua_State *lstate)
```
```suggestion
bool nlua_is_deferred_safe(lua_State *lstate)
```

Usually `_cb` suffix means the thing is itself a callback, no?
